src/System.Private.CoreLib/shared/System/IO/BinaryReader.cs

```diff
@@ -649,6 +597,46 @@ public virtual byte[] ReadBytes(int count)
             return result;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ReadOnlySpan<byte> InternalRead(int numBytes)
+        {
+            Debug.Assert(numBytes >= 2 && numBytes <= 16, "value of 1 should use ReadByte. value > 16 requires to change the minimal _buffer size");
+
+            if (_isMemoryStream)
+            {
+                // no need to check if _stream == null as we will never have null _stream when _isMemoryStream = true
+
+                // read directly from MemoryStream buffer
+                MemoryStream mStream = _stream as MemoryStream;
+                Debug.Assert(mStream != null, "_stream as MemoryStream != null");
+
+                return mStream.InternalReadSpan(numBytes);
+            }
+            else
+            {
+                if (_stream == null)
+                {
+                    throw Error.GetFileNotOpen();
+                }
+
+                int bytesRead = 0;
+                int n = 0;
+
+                do
+                {
+                    n = _stream.Read(_buffer, bytesRead, numBytes - bytesRead);
+                    if (n == 0)
+                    {
+                        throw Error.GetEndOfFile();
+                    }
+                    bytesRead += n;
+                } while (bytesRead < numBytes);
+
+                return _buffer;
+            }
+        }
+
+        [Obsolete("This should not be part of the exposed surface of this class. It is also not used internally anymore.", true)]
```

Comment:
Hmm, this is a breaking change, both marking this as Obsolete and not using this method anymore, as a derived type expecting this to be used will be sorely disappointed.

@jkotas, were you aware of this issue and decided it was fine due to the impact likely being minimal?

Reply:
This is @TomerWeisberg 's change. I did not have anything like this in my snippet 
https://github.com/jkotas/coreclr/commit/dc04600654d1db0d43050630be89a5fcfd606f42

Agree that this should stay is it was. I do not think there is a point in trying to obsolete it nor doing anything else about it to gain a bit of extra performance. The BinaryReader is not a top-performance abstraction by design. If somebody wants the fastest reader, they should use something else.

Reply:
I marked it as Obsolete to start this discussion. I thought it is not needed and if you will search for it in your favorite search engine you will see people asking why they need it and even some wrong assumptions about it. I don't see how derived type can safely use it without also overriding the Read* methods and basically create something completely custom. As the _buffer is private how can someone change this implementation in a way which will be useful?

Reply:
It is fine to start a discussion about issues like these, but the right place to do it is in a separate issue in corefx repo where we track all API additions, etc. 

We prefer smaller PRs that are fixing one problem. They are easier to review and get through the system.

Reply:
ok, I removed the Obsolete attribute but I still have no clue when someone should call that function or override it. Can you explain this?

Reply:
If there is a public extension point for 15+ years, people will find creative ways to use it.

I agree with you that this API is not designed well. We typically need to have a good reason to do something about poorly designed APIs (e.g. security risk or high rate of misuse), otherwise it is just not worth the pain to deal with the breaking change. We treat API removals or obsoleting the same way as API additions: File an issue with the proposal in corefx repo, e.g. https://github.com/dotnet/corefx/issues/33602 is an issue about obsoleting `String.Copy`.